### PR TITLE
Experimentation Interface

### PR DESCRIPTION
**Pull request by soumik12345**
- [ ] Add a CLI for experimenting with ShapenetCore
- [ ] Implement Inference Callback
- [ ] Integrate experimentation with W&B

### CODE REVIEW

File: train_shapenet_core.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+    "--drop_every", "-d", default=20, help="Epochs after which Learning Rate is dropped"
+)
+@click.option("--decay_factor", "-f", default=0.5, help="Learning Rate Decay Factor")
+@click.option("--epochs", "-e", default=60, help="Number of training epochs")
+@click.option(
+    "--use_baseline_model",
+    "-m",
+    is_flag=True,
+    help="Flag: Use Baseline Model or ShapenetCore Segmenbtation Model",
+)
```
Comment (sayakpaul) [inline train_shapenet_core.py:32]: Any particular advantages of using `click` compared to `argparse`?

Alternatively, you could also give `ml_collections` a try. Some examples: https://github.com/google-research/vision_transformer/tree/main/vit_jax/configs.

File: train_shapenet_core.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+    lr_scheduler = utils.StepDecay(initial_lr, drop_every, decay_factor)
+    lr_callback = callbacks.LearningRateScheduler(
+        lambda epoch: lr_scheduler(epoch), verbose=True
+    )
+    logs_dir = f'logs_{datetime.utcnow().strftime("%y%m%d-%H%M%S")}'
+    tb_callback = callbacks.TensorBoard(log_dir=logs_dir)
+    checkpoint_callback = callbacks.ModelCheckpoint(
+        filepath=os.path.join(logs_dir, "checkpoints", "model_{epoch}"),
+        save_weights_only=True,
+    )
```
Comment (sayakpaul) [inline train_shapenet_core.py:67]: Did you want to introduce a few spaces before the semantic code blocks? For example, you could group together the code relating to data loading into a block. It reads cleaner and easier that way. You can refer to any PyImageSearch blog post to get an idea of what I am suggesting. It feels like I am reading something without any punctuation.

Comment (soumik12345) [reply train_shapenet_core.py]: @sayakpaul tried grouping semantic code blocks in a group. Can you please take a look?

File: README.md [hunk 3]
```diff
@@ -15,14 +15,17 @@
...
+  --experiment_configs    Experiment configs (configs/shapenetcore.py)
+  --object_category       ShapenetCore object category (DEFAULT: 'Airplane')
+  --in_memory             Flag: Use In-memory dataloader (DEFAULT: True)
+  --batch_size            Batch Size (DEFAULT: 32)
+  --num_points            Number of points to be sampled from a given point cloud (DEFAULT: 1024)
+  --initial_lr FLOAT      Initial Learning Rate (DEFAULT: 1e-3)
+  --drop_every            Epochs after which Learning Rate is dropped (DEFAULT: 20)
+  --decay_factor          Learning Rate Decay Factor (DEFAULT: 0.5)
+  --epochs                Number of training epochs (DEFAULT: 50)
+  --use_baseline_model    Flag: Use Baseline Model or ShapenetCore Segmenbtation Model (DEFAULT: False)
```
Comment (sayakpaul) [inline README.md:30]: "Segmentation".

Also, I think our baseline is actually a segmentation model inspired by the PointNet paper itself, just that it's not suitable for part segmentation and needs a little bit of modification for that. I think we should make that explicitly clear in the flag.

File: README.md [hunk 4]
```diff
@@ -15,14 +15,17 @@
...
-  --epochs                    Number of training epochs
-  --use_baseline_model        Flag: Use Baseline Model or ShapenetCore Segmenbtation Model
+  --project_name          Project Name (DEFAULT: pointnet_shapenet_core)
+  --experiment_name       Experiment Name (DEFAULT: shapenet_core_experiment)
+  --wandb_api_key         W&B API Key (OPTIONAL)
+  --experiment_configs    Experiment configs (configs/shapenetcore.py)
+  --object_category       ShapenetCore object category (DEFAULT: 'Airplane')
+  --in_memory             Flag: Use In-memory dataloader (DEFAULT: True)
+  --batch_size            Batch Size (DEFAULT: 32)
+  --num_points            Number of points to be sampled from a given point cloud (DEFAULT: 1024)
```
Comment (sayakpaul) [inline README.md:25]: Should this be 2048?

File: README.md [hunk 5]
```diff
@@ -15,14 +15,17 @@
...
-  --initial_lr FLOAT          Initial Learning Rate
-  --drop_every                Epochs after which Learning Rate is dropped
-  --decay_factor              Learning Rate Decay Factor
-  --epochs                    Number of training epochs
-  --use_baseline_model        Flag: Use Baseline Model or ShapenetCore Segmenbtation Model
+  --project_name          Project Name (DEFAULT: pointnet_shapenet_core)
+  --experiment_name       Experiment Name (DEFAULT: shapenet_core_experiment)
+  --wandb_api_key         W&B API Key (OPTIONAL)
+  --experiment_configs    Experiment configs (configs/shapenetcore.py)
+  --object_category       ShapenetCore object category (DEFAULT: 'Airplane')
```
Comment (sayakpaul) [inline README.md:22]: So, currently, we are only supporting model training per-category basis, correct?

File: README.md [hunk 6]
```diff
@@ -15,14 +15,17 @@
...
-  --object_category           ShapenetCore object category
-  --in_memory                 Flag: Use In-memory dataloader
-  --batch_size                Batch Size
-  --num_points                Number of points to be sampled from a given point cloud
-  --initial_lr FLOAT          Initial Learning Rate
-  --drop_every                Epochs after which Learning Rate is dropped
-  --decay_factor              Learning Rate Decay Factor
-  --epochs                    Number of training epochs
-  --use_baseline_model        Flag: Use Baseline Model or ShapenetCore Segmenbtation Model
+  --project_name          Project Name (DEFAULT: pointnet_shapenet_core)
```
Comment (sayakpaul) [inline README.md:18]: `wandb_project_name`

File: point_seg/utils.py [hunk 7]
```diff
@@ -25,8 +26,20 @@ def __init__(self, initial_lr: float, drop_every: int, decay_factor: float) -> N
...
+def init_wandb(project_name, experiment_name, wandb_api_key):
+    """Initialize Wandb
+    Args:
+        project_name: project name on Wandb
+        experiment_name: experiment name on Wandb
+        wandb_api_key: Wandb API Key
+    """
+    if project_name is not None and experiment_name is not None:
+        os.environ["WANDB_API_KEY"] = wandb_api_key
+        wandb.init(project=project_name, name=experiment_name, sync_tensorboard=True)
```
Comment (sayakpaul) [inline point_seg/utils.py:45]: Since we are not using any specific TensorBoard-related capabilities yet, I think we can discard TensorBoard altogether and let `wandb` take care of the logging.

**Review by sayakpaul — CHANGES_REQUESTED** [review]

File: configs/shapenetcore.py [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+def get_config() -> ml_collections.ConfigDict:
+    config = ml_collections.ConfigDict()
+    config.object_category = "Airplane"  # Shapenet Category
+    config.in_memory = True  # Flag: Use In-memory dataloader
+    config.batch_size = 32  # Batch Size
+    config.num_points = 1024  # Number of points to be sampled from a given point cloud
+    config.initial_lr = 1e-3  # Initial Learning Rate
+    config.drop_every = 20  # Epochs after which Learning Rate is dropped
+    config.decay_factor = 0.5  # Learning Rate Decay Factor
+    config.use_baseline_model = (
```
Comment (sayakpaul) [inline configs/shapenetcore.py:13]: Why are we making it a set?

File: configs/shapenetcore.py [hunk 9]
```diff
@@ -0,0 +1,17 @@
+import ml_collections
+
+
+def get_config() -> ml_collections.ConfigDict:
```
Comment (sayakpaul) [inline configs/shapenetcore.py:4]: I'm not sure how this configuration is actually being used in the trainer script? 

Oh, I see it in here: https://github.com/google/ml_collections#config-flags.

File: train_shapenet_core.py [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+flags.DEFINE_string("experiment_name", "shapenet_core_experiment", "Experiment Name")
+flags.DEFINE_string("wandb_api_key", None, "Wandb API Key")
+config_flags.DEFINE_config_file("experiment_configs")
+
+
+def main(_):
+
+    # Initialize W&B
+    if FLAGS.wandb_api_key is not None:
+        utils.init_wandb(FLAGS.project_name, FLAGS.experiment_name, FLAGS.wandb_api_key)
```
Comment (sayakpaul) [inline train_shapenet_core.py:25]: Let's pass the configs to `wandb.init(config=...)` as well. Note that if we are passing a `ConfigDict` here, we should call `to_dict()` before passing it to the `config` argument of `wandb.init()`.

File: train_shapenet_core.py [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+from point_seg import ShapeNetCoreLoaderInMemory, ShapeNetCoreLoader
+from point_seg import models, utils
+
+
+FLAGS = flags.FLAGS
+flags.DEFINE_string("project_name", "pointnet_shapenet_core", "Project Name")
+flags.DEFINE_string("experiment_name", "shapenet_core_experiment", "Experiment Name")
+flags.DEFINE_string("wandb_api_key", None, "Wandb API Key")
+config_flags.DEFINE_config_file("experiment_configs")
+
```
Comment (sayakpaul) [inline train_shapenet_core.py:19]: Let's add a usage here. Something like `python train_shapenet_core.py -- ...`.

File: README.md (hunk 5 above)
Comment (soumik12345) [reply README.md:22]: Correct, this is because the Shapenet Core categories have different sets of labels.

File: configs/shapenetcore.py [hunk 12]
```diff
@@ -10,8 +10,6 @@ def get_config() -> ml_collections.ConfigDict:
     config.initial_lr = 1e-3  # Initial Learning Rate
     config.drop_every = 20  # Epochs after which Learning Rate is dropped
     config.decay_factor = 0.5  # Learning Rate Decay Factor
-    config.use_baseline_model = (
-        False  # Flag: Use Baseline Model or ShapenetCore Segmenbtation Model
-    )
+    config.use_baseline_model = False # Flag: Use Baseline Model or ShapenetCore Segmenbtation Model
```
Comment (sayakpaul) [inline configs/shapenetcore.py:13]: Segmenbtation -> Segmentation

**Review by sayakpaul — APPROVED** [review]
Just one minor comment.
